salt/minion.py

```diff
@@ -587,8 +587,21 @@ def eval_master(self,
                 opts.update(prep_ip_port(opts))
                 opts.update(resolve_dns(opts))
                 try:
-                    pub_channel = salt.transport.client.AsyncPubChannel.factory(self.opts, **factory_kwargs)
-                    yield pub_channel.connect()
+                    if self.opts['transport'] == 'detect':
+                        detect_connect = False
+                        self.opts['detect_mode'] = True
```

Comment:
`detect_mode` means "`transport` was originally set to `detect`, but we can't leave it like that if we're trying to connect, but we still need to remember that, so let's store it somewhere else".